### ISSUE TITLE
merge from branch not working anymore on 0.4.0

### ISSUE THREAD

**Issue by djobet** [ingydotnet/git-subrepo#317]
Hello,

I'm trying to "sell" subrepos to move from an existing mercurial setup (+ submodules) to a git setup.

Our setup has several applications using a common framework.
We have 3 remote sites (Europe, Asia, America).
One of our application has been heavily modified (tailored locally) so that now it's more or less 3 different applications still having the same repo and living in 3 different branches which are considered to be the local master.
We now want to check them out at the same time in the container.

To summarize
- git repo "common" (branch master)
- git repo "app1" (branch master, asia, america)

Target
git container with directory
- common
- europe (using subrepo app1 branch master)
- asia (using subrepo app1 branch asia)
- america (using subrepo app1 branch america)

The idea being that all 3 applications are using the same dependency to common/master and so, when refactoring common, we want to be able to bring in changes to europe/asia/america at the same time.

With version 0.3.0, I'm able to do
`git subrepo pull europe -b america`

With version 0.4.0, I cannot do it anymore
`git-subrepo: Local repository does not contain XXXXXXX. Try to 'git subrepo fetch america' of add the '-F' flag to always fetch the latest content.`

Fetching does not resolve the error, -F arg is rejected as not being valid with the pull operation.
Do you think this is an operation you want to support ? (this works with subtree)

There is also something I don't like with 0.3.0 and the way the history get integrated in the source repo for app1 : I only see an "administrative" commit when I do a merge instead of seeing the original merged commit + eventually a merge commit.

ie : with subrepos 0.3.1

    commit 7d722f4aac609008658640a3196d64dc174f05e3
    Author: user <user@XXXX>
    Date:   Fri Nov 10 15:24:58 2017+0000
    git subrepo commit app1
    Change-Id: Id58a9a1aa27f30c272b2d7dce31698d1f78208d6
    subrepo:                                                                                                  
      subdir:   "app1"
      merged:   "6e94a8d"
    upstream:
      origin:   "../app1.git"
      branch:   "master"
      commit:   "7a795cc"
    git-subrepo:
      version:  "0.3.1"
      origin:   "https://github.com/ingydotnet/git-subrepo"
      commit:   "a7ee886"

With subtrees

    commit 635884c8100d116b536f8c47b7466eec06d9ec4a
    Merge: d942867 06128b3
    Author: user <user@XXXX>
    Date:   Thu Nov 9 08:38:59 2017 +0000

    Merged app1/america in app1/master
    
    Change-Id: I773bcb168b07fcf82f3750bcd1981df649af5409

    commit 06128b3b5605ef96521e8678d467ff7e7b2eeb1c
    Author: user <user@XXXX>
    Date:   Thu Nov 9 08:28:33 2017 +0000

    2nd commit in app1 on branch america
    
    Change-Id: Icbf5bc2895613609595984bf0a5be90dfbfc76ae

Last question, do you plan to support cherry-pick ?

**Comment by ingydotnet** [comment]
@djobet Look into `git subrepo clone -F`. I think it might do what you want.

**Comment by djobet** [comment]
@ingydotnet I've given it a try, here's the output. Any advice ?

    $ git subrepo version
    git-subrepo Version: 0.4.0
    ....

    $ git subrepo clone -F -b america

    git-subrepo: Invalid option '--fetch' for 'clone'.

      Usage: git subrepo clone <repository> [<subdir>] [-b <branch>] [-f] [-m <msg>] [-e] [--method <merge|rebase>]
    ...

**Comment by grimmySwe** [comment]
@djobet If you want to take a specific snaphot of a subrepo into your current repo:
```git subrepo clone [--force] -b branch subrepo```

-F (capital f) flag is for fetching data first and can be used with branch/commit/status commands.

In your case I assume that you want to get the difference from america and merge that into your current status (europe?). Problem with that is that in the current .gitrepo file 0.4.0 you will reference a commit on your current subrepo branch (i.e. europe) that doesn't exist on the other branch. So subrepo doesn't know how to combine the two. In 0.3.0 we traversed the history to find a common commit and work from there. In theory it would be possible to do the same thing with the 0.4.0 design, but it's not implemented that way.

For both the case above and for cherry-picking you always have the option on working directly on the subrepos. So you can perform any git operation there. This will allow you to merge data between your different repos/branches and then you use subrepo to access them.

**Comment by djobet** [comment]
@grimmySwe thanks for your answer. Yes, I know I can do it directly on the subrepo, but it involves more operations. We're trying to find a setup that is as strimlined as possible for common operations, that's why I think it would be better if we can do everything from within the container.
Are merging/cherry-picking something you plan to support one day in 0.4.0 ?

**Comment by grimmySwe** [comment]
@djobet I haven't given it a lot of thought but my first reaction is that it is a little hairy to actually merge subrepo stuff within the context of the main repo. But who knows, there might be an easy way? If you have ideas, please let us know!

There is a way of merging by using the regular repo branches. But that would leave your subrepo unaware of the actual merge.

**Comment by djobet** [comment]
@grimmySwe Yes, I agree this is hairy ! The setup is not what I'd like to have but one I have to live with...

**Comment by djobet** [comment]
@grimmySwe Hello, I've tried to make some changes to the script to make my use-case work.
Unfortunately I've basically 0 knowledge of git low level commands, so I don't really know exactly what the script attempts to do and which pitfalls it tries to avoid. Is there some doc out there describing the sub-repo process ?

I've managed to somehow make the thing work with several caveats.
I've just commented out the "check for rebase" in subrepo:branch, line 1872.
It's nearly enough for git-subrepo to do the right thing. (of course, I guess it breaks other stuff otherwise this check would not be there)
I (expectedly) provoke a conflict, but I don't quite follow the "common conflict resolution workflow" : I execute all lines except the last one. Instead of executing "git subrepo commit europe", I just do a "git subrepo push europe".
- it does the right thing with the history of the subrepo itself : I see both the commit coming from the america branch + a merge commit
- but it fails to show the right state in the subrepo europe, the merged file still being in the previous state (before the merge)
- cloning again the europe directory shows the correct state as expected

I don't execute the "git subrepo commit europe" because otherwise when I "git subrepo push europe", it asks me to pull again, which provokes a 2nd conflict (the same one), which I resolve again, and then I'm able to push.

To summarize :
- "git subrepo commit europe" : has to perform a 2nd conflict resolution, directory europe in the right state at the end. subrepo's history has 3 more commits (commit from america, administrative commit, merge commit). 
- "git subrepo push europe" : leaves directory europe in the pre-merge state. subrepo's history has 2 more commits (commit from america, merge commit)

Is there any insight you can get from this experiment ?
Any advice on more experimentation ?

**Comment by grimmySwe** [comment]
@djobet Thanks for taking the time to investigate this!

Unfortunately there are no updated documentation on how it works. I have been waiting for the release of 0.4.0 to update the wiki pages.

A summary:
When you pull in changes from a subrepo we use the information stored in the .gitrepo file to recreate a virtual structure on the changes in the subrepo. We "simply" use the commit value in .gitrepo as references so that for each commit in the main repo we can see where it fits in the subrepo history. This allows us to quite easily use regular git commands to merge the two together.

The problem with your use case is that the commit in .gitrepo is unknown at the subrepo branch. So we don't know how to recreate the history. And if we should fetch the data from another branch, we can't record that fact in a good way. Because we only assume that we track one branch, and the commit we store should be on that branch. If we just keep the old commit, what would happen if we do the same operation again? If we update the commit, next time we do pull we would sit with a commit unknown to the tracking branch.

I have discussed the problem with my colleague for half an hour or so, but we can't see any viable solution to this. Currently every path seems to end up in a mess somehow. Mainly because you don't want to do thing in one repo, unknown to the other repo.

I guess this want help you very much as you are having a use case that isn't optimal for the current subrepo functionality. For me personally I would choose to do the actual subrepo merges in the subrepo. That will save you a lot of problems and make sure that you can use the full force of git.

Another approach would be to have separate branches in the main repo, one for each subrepo branch. Then perform the merge in main repo scope, BUT keeping the .gitrepo intact. Then it will look as you simply introduce new changes. You will loose the subrepo structure as the branches there will not know about the operations and if you ever want to merge the subrepo branches for real you could end up with a lot of conflicts.

**Comment by grimmySwe** [comment]
@djobet Issue #328 have some suggested solutions to this problem. Closing this now.
